template/template.go

```diff
@@ -296,3 +308,238 @@ func (e *executer) resolveOptions(options []ast.Option) (map[string]any, error)
+			return nil, fmt.Errorf("%w '%s'", ErrUnknownFunction, function.Identifier.Name)
+		}
+
+		opts, err := e.resolveOptions(function.Options)
+		if err != nil {
+			return nil, fmt.Errorf("resolve options: %w", err)
+		}
+
+		input, err := e.resolveValue(selectExpr.Operand)
+		if err != nil {
+			return nil, fmt.Errorf("resolve value: %w", err)
+		}
+
+		rslt, err := registryF.Match(input, opts)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %s", ErrSelection, err.Error())
+		}
+
+		res = append(res, rslt)
+	}
+
+	return res, nil
+}
+
+func (e *executer) resolvePreferences(m ast.Matcher, res []any) [][]string {
+	// Step 2: Resolve Preferences
+	pref := make([][]string, 0, len(res))
+
+	for i := range res {
+		var keys []string
+
+		for _, variant := range m.Variants {
+			for _, vKey := range variant.Keys {
+				if vKey.String() != "*" {
+					switch key := vKey.(type) {
+					case ast.CatchAllKey:
+						continue
+					case ast.NameLiteral:
+						keys = append(keys, string(key))
+					case ast.QuotedLiteral:
+						keys = append(keys, string(key))
+					case ast.NumberLiteral:
+						keys = append(keys, fmt.Sprint(key))
+					}
+				}
+			}
+		}
+
+		rv := res[i]
+
+		matches := matchSelectorKeys(rv, keys)
+		pref = append(pref, matches)
+	}
+
+	return pref
+}
+
+func (e *executer) filterVariants(m ast.Matcher, pref [][]string) []ast.Variant {
+	// Step 3: Filter Variants
+	var filteredVariants []ast.Variant
+
+	for _, variant := range m.Variants {
+		matchesAllSelectors := true
+
+		for i, keyOrder := range pref {
+			key := variant.Keys[i]
+			if key.String() == "*" {
+				continue
+			}
+
+			var ks string
+
+			switch key := key.(type) {
+			case ast.CatchAllKey:
+				continue
+			case ast.NameLiteral:
+				ks = string(key)
+			case ast.QuotedLiteral:
+				ks = string(key)
+			case ast.NumberLiteral:
+				ks = fmt.Sprint(key)
+			}
+
+			if !slices.Contains(keyOrder, ks) {
+				matchesAllSelectors = false
+				break
+			}
+		}
+
+		if matchesAllSelectors {
+			filteredVariants = append(filteredVariants, variant)
+		}
+	}
+
+	return filteredVariants
+}
+
+func (e *executer) sortVariants(filteredVariants []ast.Variant, pref [][]string) []SortableVariant {
+	// Step 4: Sort Variants
+	sortable := make([]SortableVariant, 0, len(filteredVariants))
+
+	for _, variant := range filteredVariants {
+		sortable = append(sortable, SortableVariant{Score: -1, Variant: variant})
+	}
+
+	for i := len(pref) - 1; i >= 0; i-- {
+		matches := pref[i]
+
+		for tupleIndex, tuple := range sortable {
+			key := tuple.Variant.Keys[i]
+			currentScore := len(matches)
+
+			if key.String() != "*" {
+				var ks string
+
+				switch key := key.(type) {
+				case ast.CatchAllKey:
+					continue
+				case ast.NameLiteral:
+					ks = string(key)
+				case ast.QuotedLiteral:
+					ks = string(key)
+				case ast.NumberLiteral:
+					ks = fmt.Sprint(key)
+				}
+
+				if position := findPosition(ks, matches); position != -1 {
+					currentScore = position
+				}
```

Comment:
is it crucial that score can be negative ? 🤔 
```suggestion
currentScore :=  findPosition(ks, matches)
```